cpp/src/umap/knn_graph/algo.cuh

```diff
@@ -101,6 +101,16 @@ inline void launcher(const raft::handle_t& handle,
                      const ML::UMAPParams* params,
                      cudaStream_t stream)
 {
+  size_t free_mem, total_mem;
+  cudaMemGetInfo(&free_mem, &total_mem);
```

Comment:
This is, unfortunately, not a reliable way to query and compute the memory available for batching. It also won't work well with rmm pools.

The challenge here is that the total memory space can quickly become very fragmented, which prevents the reporter free memory from being allocated in a single chunk. Further, if 90% of the memory is allocated to a pool, this is going to report almost no memory available even though the pool could be largely empty. 

A better way to approach this problem will be to use the limiting memory resource once irs available. This new resource will be applied on top of the workspace resource and will provide a safe and consistent mechanism for us to handle implicit batching on the users behalf.

Reply:
@cjnolet, the current approach of this PR was to first compute the memory requirements, and then adjust the index batch size if necessary. I understand that getting the free available memory directly from CUDA API is not a good solution as a RMM pool that could/should be used might have already been pre-allocated.

The `limiting_resource_adaptor` might provide a way to limit downstream memory usage of an upstream `device_memory_resource`. However, I do not see how this will help solving this issue. The issue is that a complete sparse operation consumes too much memory at once and should be split in multiple operations through batching instead.

Could you give more precision as to how it _**will provide a safe and consistent mechanism for us to handle implicit batching on the users behalf in a way that also plays nicely with RMM pools**_?